File: packages/sdk/arc-py/examples/governed_hello.py
from __future__ import annotations

import json
import os
import sys

import httpx

from arc import ArcClient, ReceiptQueryClient


def require_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise SystemExit(f"missing required environment variable: {name}")
    return value


def session_capability_id(base_url: str, auth_token: str, session_id: str) -> str:
    response = httpx.get(
        f"{base_url}/admin/sessions/{session_id}/trust",
        headers={"Authorization": f"Bearer {auth_token}"},
        timeout=5.0,
    )
    response.raise_for_status()
    payload = response.json()
    capabilities = payload.get("capabilities") or []
    capability_id = capabilities[0].get("capabilityId") if capabilities else None
    if not isinstance(capability_id, str) or not capability_id:
        raise RuntimeError("session trust endpoint did not return an active capability id")
    return capability_id


def main() -> None:
    base_url = require_env("ARC_BASE_URL")
    control_url = os.environ.get("ARC_CONTROL_URL", base_url)
    auth_token = require_env("ARC_AUTH_TOKEN")
    message = sys.argv[1] if len(sys.argv) > 1 else "hello from the Python SDK"

    client = ArcClient.with_static_bearer(base_url, auth_token)
    session = client.initialize(
        client_info={"name": "arc-sdk/examples/python", "version": "1.0.0"}
    )

    try:
        tools_result = session.list_tools()
        tools = tools_result.get("result", {}).get("tools", [])
        capability_id = session_capability_id(base_url, auth_token, session.session_id)
        tool_result = session.call_tool("echo_text", {"message": message}).get("result", {})
        receipts = ReceiptQueryClient(control_url, auth_token).query(
            {"capabilityId": capability_id, "limit": 10}
        )
        receipt_list = receipts.get("receipts", [])
        if not receipt_list:
            raise RuntimeError("receipt query did not return the governed tool receipt")
        receipt = receipt_list[-1]

        print(
            json.dumps(
                {
                    "sessionId": session.session_id,
                    "capabilityId": capability_id,
                    "toolNames": [tool["name"] for tool in tools],
                    "echo": tool_result.get("structuredContent", {}).get("echo"),
                    "receiptId": receipt.get("id"),
                    "receiptDecision": receipt.get("decision"),
                },
                indent=2,
            )
        )
    finally:
        session.close()


if __name__ == "__main__":
    main()

File: packages/sdk/arc-ts/examples/governed_hello.ts
import { ArcClient, ReceiptQueryClient } from "../src/index.ts";

function requireEnv(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`missing required environment variable: ${name}`);
  }
  return value;
}

async function sessionCapabilityId(
  baseUrl: string,
  authToken: string,
  sessionId: string,
): Promise<string> {
  const response = await fetch(`${baseUrl}/admin/sessions/${sessionId}/trust`, {
    headers: { Authorization: `Bearer ${authToken}` },
  });
  if (!response.ok) {
    throw new Error(`session trust query failed with HTTP ${response.status}`);
  }
  const payload = (await response.json()) as {
    capabilities?: Array<{ capabilityId?: string }>;
  };
  const capabilityId = payload.capabilities?.[0]?.capabilityId;
  if (!capabilityId) {
    throw new Error("session trust endpoint did not return an active capability id");
  }
  return capabilityId;
}

const baseUrl = requireEnv("ARC_BASE_URL");
const controlUrl = process.env.ARC_CONTROL_URL ?? baseUrl;
const authToken = requireEnv("ARC_AUTH_TOKEN");
const message = process.argv[2] ?? "hello from the TypeScript SDK";

const client = ArcClient.withStaticBearer(baseUrl, authToken);
const session = await client.initialize({
  clientInfo: {
    name: "@arc-protocol/sdk/examples/typescript",
    version: "1.0.0",
  },
});

try {
  const tools = (await session.listTools()) as { tools?: Array<{ name: string }> };
  const capabilityId = await sessionCapabilityId(baseUrl, authToken, session.sessionId);
  const toolResult = (await session.callTool("echo_text", { message })) as {
    content?: Array<{ text?: string }>;
    structuredContent?: { echo?: string };
  };
  const receipts = await new ReceiptQueryClient(controlUrl, authToken).query({
    capabilityId,
    limit: 10,
  });
  const receipt = receipts.receipts.at(-1);
  if (!receipt) {
    throw new Error("receipt query did not return the governed tool receipt");
  }

  console.log(
    JSON.stringify(
      {
        sessionId: session.sessionId,
        capabilityId,
        toolNames: tools.tools?.map((tool) => tool.name) ?? [],
        echo:
          toolResult.structuredContent?.echo ??
          toolResult.content?.[0]?.text ??
          null,
        receiptId: receipt.id,
        receiptDecision: receipt.decision,
      },
      null,
      2,
    ),
  );
} finally {
  await session.close();
}
